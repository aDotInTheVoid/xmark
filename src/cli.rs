use clap::Clap;
use eyre::Result;
use std::path::PathBuf;

#[derive(Debug, Default, Clone)]
pub struct Args {
    pub dir: PathBuf,
    pub create: bool,
    pub templates: Option<PathBuf>,
}

impl Args {
    pub fn parse() -> Result<Self> {
        let ArgsInner {
            mut dir,
            create,
            mut templates,
        } = ArgsInner::parse();
        dir = dir.canonicalize()?;
        templates = match templates {
            Some(p) => Some(p.canonicalize()?),
            None => None,
        };
        Ok(Args {
            dir,
            create,
            templates,
        })
    }
}

#[derive(Clap, Debug)]
struct ArgsInner {
    #[clap(short, long, default_value = ".")]
    pub dir: PathBuf,
    #[clap(long)]
    pub create: bool,
    #[clap(short, long)]
    pub templates: Option<PathBuf>,
}
