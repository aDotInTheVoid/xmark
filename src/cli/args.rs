use clap::Clap;
use eyre::Result;
use tracing::instrument;

#[derive(Clap, Debug)]
pub(super) struct Args {
    #[clap(subcommand)]
    subcmd: SubCommand,
}

#[derive(Clap, Debug)]
pub(super) enum SubCommand {
    Build(super::cmd::build::Args), // TODO: init, clean, watch, serve
}

impl Args {
    #[instrument]
    pub fn run(self) -> Result<()> {
        match self.subcmd {
            SubCommand::Build(c) => c.run(),
        }
    }
}
