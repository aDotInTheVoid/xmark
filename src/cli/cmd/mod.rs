//! Individual subcommands.
pub mod build;