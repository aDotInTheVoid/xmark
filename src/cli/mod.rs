// SPDX-License-Identifier: GPL-3.0-only
use clap::Clap;
use eyre::Result;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use tracing::instrument;

pub mod config;
pub mod summary;

mod create_missing;

#[derive(Debug, Default, Clone, Serialize, Deserialize)]
pub struct Args {
    pub dir: PathBuf,
    pub create: bool,
    pub templates: Option<PathBuf>,
}

impl Args {
    #[instrument]
    pub fn parse() -> Result<Self> {
        Self::parse_from(ArgsInner::parse())
    }

    #[instrument]
    fn parse_from(inner: ArgsInner) -> Result<Self> {
        let ArgsInner {
            mut dir,
            create,
            mut templates,
        } = inner;
        dir = dir.canonicalize()?;
        templates = match templates {
            Some(p) => Some(p.canonicalize()?),
            None => None,
        };
        Ok(Args {
            dir,
            create,
            templates,
        })
    }
}

#[instrument]
pub fn init() -> Result<()> {
    color_eyre::install()?;

    use tracing_error::ErrorLayer;
    use tracing_subscriber::prelude::*;
    use tracing_subscriber::{fmt, EnvFilter};

    let fmt_layer = fmt::layer().with_target(false);
    let filter_layer = EnvFilter::try_from_default_env()
        .or_else(|_| EnvFilter::try_new("info"))
        .unwrap();

    tracing_subscriber::registry()
        .with(filter_layer)
        .with(fmt_layer)
        .with(ErrorLayer::default())
        .init();

    Ok(())
}

#[derive(Clap, Debug)]
struct ArgsInner {
    #[clap(short, long, default_value = ".")]
    pub dir: PathBuf,
    #[clap(long)]
    pub create: bool,
    #[clap(short, long)]
    pub templates: Option<PathBuf>,
}

#[cfg(test)]
mod tests {
    use insta::{assert_yaml_snapshot, dynamic_redaction};

    use crate::test_utils::manifest_dir_redacter;

    use super::*;
    #[test]
    fn canonicalize_cwd_dir() {
        let args = ArgsInner {
            dir: ".".into(),
            create: true,
            templates: None,
        };
        let args = Args::parse_from(args).unwrap();
        // Yaml uses ~ for null
        assert_yaml_snapshot!(args, {
            ".*" => dynamic_redaction(manifest_dir_redacter)
        });
    }

    #[test]
    fn canonicalize_rel_templ() {
        let args = ArgsInner {
            dir: ".".into(),
            create: true,
            // This needs to exist, as `.canonicalize` will read, in case it's a symblink
            templates: Some("./www".into()),
        };
        let args = Args::parse_from(args).unwrap();
        assert_yaml_snapshot!(args, {
            ".*" => dynamic_redaction(manifest_dir_redacter)
        });
    }

    #[test]
    fn absolute() {
        let args = ArgsInner {
            // This should exist on most non windows systems
            // TODO: win
            dir: "/usr".into(),
            create: false,
            templates: None,
        };
        let args = Args::parse_from(args).unwrap();
        assert_yaml_snapshot!(args, {
            ".*" => dynamic_redaction(manifest_dir_redacter)
        });
    }
}
