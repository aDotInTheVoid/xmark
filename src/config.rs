use std::fs;
use std::path::PathBuf;

use eyre::{Result, WrapErr};
use serde::{Deserialize, Serialize};

use crate::cli;
use crate::create_missing::create_missing;
use crate::summary::{parse_summary, Chapter, Summary};

/// The Config as represented in the global xmark.toml
#[derive(Clone, Debug, Hash, Serialize, Deserialize)]
pub struct GlobalConfigRepr {
    pub books: Vec<String>,
}

/// The config as usable for the programm
#[derive(Clone, Debug, Hash, PartialEq, Serialize, Deserialize)]
pub struct GlobalConf {
    pub books: Vec<BookConf>,
}

#[derive(Clone, Debug, Hash, PartialEq, Serialize, Deserialize)]
pub struct BookConf {
    pub location: PathBuf,
    pub slug: String,
    pub summary: Summary,
}

pub fn load(args: &cli::Args) -> Result<GlobalConf> {
    let conf = fs::read_to_string(args.dir.clone().join("xmark.toml"))
        .with_context(|| "Couldn't find xmark.toml")?;
    let conf: GlobalConfigRepr = toml::from_str(&conf)?;

    hydrate(conf, &args)
}

// Convert the disk format to a usable form
pub fn hydrate(gcr: GlobalConfigRepr, args: &cli::Args) -> Result<GlobalConf> {
    Ok(GlobalConf {
        books: gcr
            .books
            .iter()
            .map(|name| {
                let location = args.dir.join(name);
                let summary_location = location.join("SUMMARY.md");
                let mut summary = parse_summary(
                    &fs::read_to_string(&summary_location)
                        .wrap_err_with(|| format!("Couldn't open {:?}", summary_location))?,
                )?;

                if args.create {
                    create_missing(&location, &summary)?;
                }

                let fix_chap_loc = |chap: &mut Chapter| {
                    if let Some(loc) = chap.location.as_deref() {
                        chap.location = Some(location.join(loc));
                    }
                };

                summary.prefix_chapters.iter_mut().for_each(fix_chap_loc);

                summary.suffix_chapters.iter_mut().for_each(fix_chap_loc);

                summary.numbered_chapters.iter_mut().for_each(|chap| {
                    chap.map_mut(fix_chap_loc);
                });

                Ok(BookConf {
                    location,
                    summary,
                    slug: name.to_owned(),
                })
            })
            .collect::<Result<_>>()?,
    })
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;

    #[test]
    fn de_gloal_config() {
        let inp = "books = [
            'book-1',
            'book-2',
            'book-3',
            ]
";
        let conf: GlobalConfigRepr = toml::from_str(inp).unwrap();
        assert_eq!(conf.books, vec!["book-1", "book-2", "book-3"]);
    }

    #[test]
    fn hydrate_basic() {
        let args = cli::Args {
            dir: "/home/etc/bax".into(),
            create: false,
        };
        let gcr = GlobalConfigRepr { books: vec![] };
        let gc = GlobalConf { books: vec![] };
        assert_eq!(hydrate(gcr, &args).unwrap(), gc);
    }

    #[test]
    fn hydrate_dummy() {
        // I don't know why I need the anotation
        let redacter = |mut val, _: insta::internals::ContentPath| {
            // TODO: Dont do insta crimes.
            // This is in #[doc(Hidden)] internals.
            while let insta::internals::Content::Some(some) = val {
                val = *some;
            }

            if let insta::internals::Content::String(s) = val {
                s.replace(dbg!(env!("CARGO_MANIFEST_DIR")), "BASEDIR")
            } else {
                "".into()
            }
        };

        let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("dummy-book");
        let args = cli::Args { dir, create: false };
        let conf = load(&args).unwrap();
        assert_yaml_snapshot!(conf, {
            ".location" => insta::dynamic_redaction(redacter),
            ".**.location" => insta::dynamic_redaction(redacter),
        });
    }
}
