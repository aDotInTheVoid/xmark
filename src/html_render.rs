use std::fs;
use std::io::Write;
use std::path::Path;

use eyre::{Context, Result};
use pulldown_cmark::{html, Options, Parser};

use crate::cli;
use crate::config::BookConf;
use crate::summary::Chapter;

pub fn render(book: BookConf, args: &cli::Args) -> Result<()> {
    let out_dir = args.dir.clone().join("_out").join("html");
    fs::create_dir_all(&out_dir)?;

    for i in book.summary.prefix_chapters {
        render_chap_io(&i, &out_dir, &args.dir)?;
    }
    for i in book.summary.suffix_chapters {
        render_chap_io(&i, &out_dir, &args.dir)?;
    }
    for i in book.summary.numbered_chapters {
        i.try_map(|chap| render_chap_io(chap, &out_dir, &args.dir))?;
    }

    Ok(())
}

fn render_chap_io(chapter: &Chapter, build_dir: &Path, base_dir: &Path) -> Result<()> {
    // Not a draft
    if let Some(ref loc) = chapter.location {
        let content = fs::read_to_string(loc)?;
        let html = render_chap(&content);

        let mut path = build_dir.join(loc.strip_prefix(base_dir).expect("Unreachble"));

        path.set_extension("html");

        fs::create_dir_all(path.parent().unwrap())?;

        let mut file =
            fs::File::create(&path).wrap_err_with(|| format!("Failed to create {:?}", &path))?;
        file.write_all(html.as_bytes())?;
    }

    Ok(())
}

// Pure inner for testing
fn render_chap(content: &str) -> String {
    let opts = Options::all();
    let parser = Parser::new_ext(content, opts);
    let mut out = String::new();
    html::push_html(&mut out, parser);
    out
}

#[cfg(test)]
mod tests {
    use std::collections::{BTreeSet};
    use std::path::PathBuf;

    use assert_fs::prelude::*;
    use insta::{assert_yaml_snapshot, dynamic_redaction};

    use crate::{cli, config, html_render};

    #[test]
    fn dummy_e2e() {
        let temp = assert_fs::TempDir::new().unwrap();
        temp.copy_from(
            PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("dummy-book"),
            &["xmark.toml", "book-*/**"],
        )
        .unwrap();

        let args = cli::Args {
            create: false,
            dir: temp.path().to_owned(),
        };
        let conf = config::load(&args).unwrap();
        for book in conf.books {
            html_render::render(book, &args).unwrap()
        }


        // BTree so it's in order.
        let paths: BTreeSet<_> = ignore::Walk::new(temp.path())
            .filter_map(|x| x.ok().map(|y| y.into_path()))
            .map(|x| x.into_os_string().into_string().unwrap())
            .collect();

        assert_yaml_snapshot!(paths, {"[]" => dynamic_redaction(move |val, _| {
            val.as_str().unwrap().replace(temp.path().as_os_str().to_str().unwrap(), "")
        })});
    }
}
