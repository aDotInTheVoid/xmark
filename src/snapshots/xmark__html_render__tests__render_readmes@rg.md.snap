---
source: src/html_render.rs
expression: out
input_file: src/render_html_tests/rg.md
---
<h2>ripgrep (rg)</h2>
<p>ripgrep is a line-oriented search tool that recursively searches your current
directory for a regex pattern. By default, ripgrep will respect your .gitignore
and automatically skip hidden files/directories and binary files. ripgrep
has first class support on Windows, macOS and Linux, with binary downloads
available for <a href="https://github.com/BurntSushi/ripgrep/releases">every release</a>.
ripgrep is similar to other popular search tools like The Silver Searcher, ack
and grep.</p>
<p><a href="https://github.com/BurntSushi/ripgrep/actions"><img src="https://github.com/BurntSushi/ripgrep/workflows/ci/badge.svg" alt="Build status" /></a>
<a href="https://crates.io/crates/ripgrep"><img src="https://img.shields.io/crates/v/ripgrep.svg" alt="Crates.io" /></a>
<a href="https://repology.org/project/ripgrep/badges"><img src="https://repology.org/badge/tiny-repos/ripgrep.svg" alt="Packaging status" /></a></p>
<p>Dual-licensed under MIT or the <a href="https://unlicense.org">UNLICENSE</a>.</p>
<h3>CHANGELOG</h3>
<p>Please see the <a href="CHANGELOG.md">CHANGELOG</a> for a release history.</p>
<h3>Documentation quick links</h3>
<ul>
<li><a href="#installation">Installation</a></li>
<li><a href="GUIDE.md">User Guide</a></li>
<li><a href="FAQ.md">Frequently Asked Questions</a></li>
<li><a href="https://docs.rs/regex/1/regex/#syntax">Regex syntax</a></li>
<li><a href="GUIDE.md#configuration-file">Configuration files</a></li>
<li><a href="FAQ.md#complete">Shell completions</a></li>
<li><a href="#building">Building</a></li>
<li><a href="#translations">Translations</a></li>
</ul>
<h3>Screenshot of search results</h3>
<p><a href="https://burntsushi.net/stuff/ripgrep1.png"><img src="https://burntsushi.net/stuff/ripgrep1.png" alt="A screenshot of a sample search with ripgrep" /></a></p>
<h3>Quick examples comparing tools</h3>
<p>This example searches the entire
<a href="https://github.com/BurntSushi/linux">Linux kernel source tree</a>
(after running <code>make defconfig &amp;&amp; make -j8</code>) for <code>[A-Z]+_SUSPEND</code>, where
all matches must be words. Timings were collected on a system with an Intel
i7-6900K 3.2 GHz.</p>
<p>Please remember that a single benchmark is never enough! See my
<a href="https://blog.burntsushi.net/ripgrep/">blog post on ripgrep</a>
for a very detailed comparison with more benchmarks and analysis.</p>
<table><thead><tr><th>Tool</th><th>Command</th><th>Line count</th><th>Time</th></tr></thead><tbody>
<tr><td>ripgrep (Unicode)</td><td><code>rg -n -w '[A-Z]+_SUSPEND'</code></td><td>452</td><td><strong>0.136s</strong></td></tr>
<tr><td><a href="https://www.kernel.org/pub/software/scm/git/docs/git-grep.html">git grep</a></td><td><code>git grep -P -n -w '[A-Z]+_SUSPEND'</code></td><td>452</td><td>0.348s</td></tr>
<tr><td><a href="https://github.com/Genivia/ugrep">ugrep (Unicode)</a></td><td><code>ugrep -r --ignore-files --no-hidden -I -w '[A-Z]+_SUSPEND'</code></td><td>452</td><td>0.506s</td></tr>
<tr><td><a href="https://github.com/ggreer/the_silver_searcher">The Silver Searcher</a></td><td><code>ag -w '[A-Z]+_SUSPEND'</code></td><td>452</td><td>0.654s</td></tr>
<tr><td><a href="https://www.kernel.org/pub/software/scm/git/docs/git-grep.html">git grep</a></td><td><code>LC_ALL=C git grep -E -n -w '[A-Z]+_SUSPEND'</code></td><td>452</td><td>1.150s</td></tr>
<tr><td><a href="https://github.com/beyondgrep/ack3">ack</a></td><td><code>ack -w '[A-Z]+_SUSPEND'</code></td><td>452</td><td>4.054s</td></tr>
<tr><td><a href="https://www.kernel.org/pub/software/scm/git/docs/git-grep.html">git grep (Unicode)</a></td><td><code>LC_ALL=en_US.UTF-8 git grep -E -n -w '[A-Z]+_SUSPEND'</code></td><td>452</td><td>4.205s</td></tr>
</tbody></table>
<p>Here’s another benchmark on the same corpus as above that disregards gitignore
files and searches with a whitelist instead. The corpus is the same as in the
previous benchmark, and the flags passed to each command ensure that they are
doing equivalent work:</p>
<table><thead><tr><th>Tool</th><th>Command</th><th>Line count</th><th>Time</th></tr></thead><tbody>
<tr><td>ripgrep</td><td><code>rg -uuu -tc -n -w '[A-Z]+_SUSPEND'</code></td><td>388</td><td><strong>0.096s</strong></td></tr>
<tr><td><a href="https://github.com/Genivia/ugrep">ugrep</a></td><td><code>ugrep -r -n --include='*.c' --include='*.h' -w '[A-Z]+_SUSPEND'</code></td><td>388</td><td>0.493s</td></tr>
<tr><td><a href="https://www.gnu.org/software/grep/">GNU grep</a></td><td><code>egrep -r -n --include='*.c' --include='*.h' -w '[A-Z]+_SUSPEND'</code></td><td>388</td><td>0.806s</td></tr>
</tbody></table>
<p>And finally, a straight-up comparison between ripgrep, ugrep and GNU grep on a
single large file cached in memory
(~13GB, <a href="http://opus.nlpl.eu/download.php?f=OpenSubtitles/v2018/mono/OpenSubtitles.raw.en.gz"><code>OpenSubtitles.raw.en.gz</code></a>):</p>
<table><thead><tr><th>Tool</th><th>Command</th><th>Line count</th><th>Time</th></tr></thead><tbody>
<tr><td>ripgrep</td><td><code>rg -w 'Sherlock [A-Z]\w+'</code></td><td>7882</td><td><strong>2.769s</strong></td></tr>
<tr><td><a href="https://github.com/Genivia/ugrep">ugrep</a></td><td><code>ugrep -w 'Sherlock [A-Z]\w+'</code></td><td>7882</td><td>6.802s</td></tr>
<tr><td><a href="https://www.gnu.org/software/grep/">GNU grep</a></td><td><code>LC_ALL=en_US.UTF-8 egrep -w 'Sherlock [A-Z]\w+'</code></td><td>7882</td><td>9.027s</td></tr>
</tbody></table>
<p>In the above benchmark, passing the <code>-n</code> flag (for showing line numbers)
increases the times to <code>3.423s</code> for ripgrep and <code>13.031s</code> for GNU grep. ugrep
times are unaffected by the presence or absence of <code>-n</code>.</p>
<h3>Why should I use ripgrep?</h3>
<ul>
<li>It can replace many use cases served by other search tools
because it contains most of their features and is generally faster. (See
<a href="FAQ.md#posix4ever">the FAQ</a> for more details on whether ripgrep can truly
replace grep.)</li>
<li>Like other tools specialized to code search, ripgrep defaults to recursive
directory search and won’t search files ignored by your
<code>.gitignore</code>/<code>.ignore</code>/<code>.rgignore</code> files. It also ignores hidden and binary
files by default. ripgrep also implements full support for <code>.gitignore</code>,
whereas there are many bugs related to that functionality in other code
search tools claiming to provide the same functionality.</li>
<li>ripgrep can search specific types of files. For example, <code>rg -tpy foo</code>
limits your search to Python files and <code>rg -Tjs foo</code> excludes JavaScript
files from your search. ripgrep can be taught about new file types with
custom matching rules.</li>
<li>ripgrep supports many features found in <code>grep</code>, such as showing the context
of search results, searching multiple patterns, highlighting matches with
color and full Unicode support. Unlike GNU grep, ripgrep stays fast while
supporting Unicode (which is always on).</li>
<li>ripgrep has optional support for switching its regex engine to use PCRE2.
Among other things, this makes it possible to use look-around and
backreferences in your patterns, which are not supported in ripgrep’s default
regex engine. PCRE2 support can be enabled with <code>-P/--pcre2</code> (use PCRE2
always) or <code>--auto-hybrid-regex</code> (use PCRE2 only if needed). An alternative
syntax is provided via the <code>--engine (default|pcre2|auto-hybrid)</code> option.</li>
<li>ripgrep supports searching files in text encodings other than UTF-8, such
as UTF-16, latin-1, GBK, EUC-JP, Shift_JIS and more. (Some support for
automatically detecting UTF-16 is provided. Other text encodings must be
specifically specified with the <code>-E/--encoding</code> flag.)</li>
<li>ripgrep supports searching files compressed in a common format (brotli,
bzip2, gzip, lz4, lzma, xz, or zstandard) with the <code>-z/--search-zip</code> flag.</li>
<li>ripgrep supports
<a href="GUIDE.md#preprocessor">arbitrary input preprocessing filters</a>
which could be PDF text extraction, less supported decompression, decrypting,
automatic encoding detection and so on.</li>
</ul>
<p>In other words, use ripgrep if you like speed, filtering by default, fewer
bugs and Unicode support.</p>
<h3>Why shouldn’t I use ripgrep?</h3>
<p>Despite initially not wanting to add every feature under the sun to ripgrep,
over time, ripgrep has grown support for most features found in other file
searching tools. This includes searching for results spanning across multiple
lines, and opt-in support for PCRE2, which provides look-around and
backreference support.</p>
<p>At this point, the primary reasons not to use ripgrep probably consist of one
or more of the following:</p>
<ul>
<li>You need a portable and ubiquitous tool. While ripgrep works on Windows,
macOS and Linux, it is not ubiquitous and it does not conform to any
standard such as POSIX. The best tool for this job is good old grep.</li>
<li>There still exists some other feature (or bug) not listed in this README that
you rely on that’s in another tool that isn’t in ripgrep.</li>
<li>There is a performance edge case where ripgrep doesn’t do well where another
tool does do well. (Please file a bug report!)</li>
<li>ripgrep isn’t possible to install on your machine or isn’t available for your
platform. (Please file a bug report!)</li>
</ul>
<h3>Is it really faster than everything else?</h3>
<p>Generally, yes. A large number of benchmarks with detailed analysis for each is
<a href="https://blog.burntsushi.net/ripgrep/">available on my blog</a>.</p>
<p>Summarizing, ripgrep is fast because:</p>
<ul>
<li>It is built on top of
<a href="https://github.com/rust-lang/regex">Rust’s regex engine</a>.
Rust’s regex engine uses finite automata, SIMD and aggressive literal
optimizations to make searching very fast. (PCRE2 support can be opted into
with the <code>-P/--pcre2</code> flag.)</li>
<li>Rust’s regex library maintains performance with full Unicode support by
building UTF-8 decoding directly into its deterministic finite automaton
engine.</li>
<li>It supports searching with either memory maps or by searching incrementally
with an intermediate buffer. The former is better for single files and the
latter is better for large directories. ripgrep chooses the best searching
strategy for you automatically.</li>
<li>Applies your ignore patterns in <code>.gitignore</code> files using a
<a href="https://docs.rs/regex/1/regex/struct.RegexSet.html"><code>RegexSet</code></a>.
That means a single file path can be matched against multiple glob patterns
simultaneously.</li>
<li>It uses a lock-free parallel recursive directory iterator, courtesy of
<a href="https://docs.rs/crossbeam"><code>crossbeam</code></a> and
<a href="https://docs.rs/ignore"><code>ignore</code></a>.</li>
</ul>
<h3>Feature comparison</h3>
<p>Andy Lester, author of <a href="https://beyondgrep.com/">ack</a>, has published an
excellent table comparing the features of ack, ag, git-grep, GNU grep and
ripgrep: https://beyondgrep.com/feature-comparison/</p>
<p>Note that ripgrep has grown a few significant new features recently that
are not yet present in Andy’s table. This includes, but is not limited to,
configuration files, passthru, support for searching compressed files,
multiline search and opt-in fancy regex support via PCRE2.</p>
<h3>Installation</h3>
<p>The binary name for ripgrep is <code>rg</code>.</p>
<p><strong><a href="https://github.com/BurntSushi/ripgrep/releases">Archives of precompiled binaries for ripgrep are available for Windows,
macOS and Linux.</a></strong> Users of
platforms not explicitly mentioned below are advised to download one of these
archives.</p>
<p>Linux binaries are static executables. Windows binaries are available either as
built with MinGW (GNU) or with Microsoft Visual C++ (MSVC). When possible,
prefer MSVC over GNU, but you’ll need to have the <a href="https://www.microsoft.com/en-us/download/details.aspx?id=48145">Microsoft VC++ 2015
redistributable</a>
installed.</p>
<p>If you’re a <strong>macOS Homebrew</strong> or a <strong>Linuxbrew</strong> user, then you can install
ripgrep from homebrew-core:</p>
<pre><code>$ brew install ripgrep
</code></pre>
<p>If you’re a <strong>MacPorts</strong> user, then you can install ripgrep from the
<a href="https://www.macports.org/ports.php?by=name&amp;substr=ripgrep">official ports</a>:</p>
<pre><code>$ sudo port install ripgrep
</code></pre>
<p>If you’re a <strong>Windows Chocolatey</strong> user, then you can install ripgrep from the
<a href="https://chocolatey.org/packages/ripgrep">official repo</a>:</p>
<pre><code>$ choco install ripgrep
</code></pre>
<p>If you’re a <strong>Windows Scoop</strong> user, then you can install ripgrep from the
<a href="https://github.com/ScoopInstaller/Main/blob/master/bucket/ripgrep.json">official bucket</a>:</p>
<pre><code>$ scoop install ripgrep
</code></pre>
<p>If you’re an <strong>Arch Linux</strong> user, then you can install ripgrep from the official repos:</p>
<pre><code>$ pacman -S ripgrep
</code></pre>
<p>If you’re a <strong>Gentoo</strong> user, you can install ripgrep from the
<a href="https://packages.gentoo.org/packages/sys-apps/ripgrep">official repo</a>:</p>
<pre><code>$ emerge sys-apps/ripgrep
</code></pre>
<p>If you’re a <strong>Fedora</strong> user, you can install ripgrep from official
repositories.</p>
<pre><code>$ sudo dnf install ripgrep
</code></pre>
<p>If you’re an <strong>openSUSE</strong> user, ripgrep is included in <strong>openSUSE Tumbleweed</strong>
and <strong>openSUSE Leap</strong> since 15.1.</p>
<pre><code>$ sudo zypper install ripgrep
</code></pre>
<p>If you’re a <strong>RHEL/CentOS 7/8</strong> user, you can install ripgrep from
<a href="https://copr.fedorainfracloud.org/coprs/carlwgeorge/ripgrep/">copr</a>:</p>
<pre><code>$ sudo yum-config-manager --add-repo=https://copr.fedorainfracloud.org/coprs/carlwgeorge/ripgrep/repo/epel-7/carlwgeorge-ripgrep-epel-7.repo
$ sudo yum install ripgrep
</code></pre>
<p>If you’re a <strong>Nix</strong> user, you can install ripgrep from
<a href="https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/text/ripgrep/default.nix">nixpkgs</a>:</p>
<pre><code>$ nix-env --install ripgrep
$ # (Or using the attribute name, which is also ripgrep.)
</code></pre>
<p>If you’re a <strong>Debian</strong> user (or a user of a Debian derivative like <strong>Ubuntu</strong>),
then ripgrep can be installed using a binary <code>.deb</code> file provided in each
<a href="https://github.com/BurntSushi/ripgrep/releases">ripgrep release</a>.</p>
<pre><code>$ curl -LO https://github.com/BurntSushi/ripgrep/releases/download/11.0.2/ripgrep_11.0.2_amd64.deb
$ sudo dpkg -i ripgrep_11.0.2_amd64.deb
</code></pre>
<p>If you run Debian Buster (currently Debian stable) or Debian sid, ripgrep is
<a href="https://tracker.debian.org/pkg/rust-ripgrep">officially maintained by Debian</a>.</p>
<pre><code>$ sudo apt-get install ripgrep
</code></pre>
<p>If you’re an <strong>Ubuntu Cosmic (18.10)</strong> (or newer) user, ripgrep is
<a href="https://launchpad.net/ubuntu/+source/rust-ripgrep">available</a> using the same
packaging as Debian:</p>
<pre><code>$ sudo apt-get install ripgrep
</code></pre>
<p>(N.B. Various snaps for ripgrep on Ubuntu are also available, but none of them
seem to work right and generate a number of very strange bug reports that I
don’t know how to fix and don’t have the time to fix. Therefore, it is no
longer a recommended installation option.)</p>
<p>If you’re a <strong>FreeBSD</strong> user, then you can install ripgrep from the
<a href="https://www.freshports.org/textproc/ripgrep/">official ports</a>:</p>
<pre><code># pkg install ripgrep
</code></pre>
<p>If you’re an <strong>OpenBSD</strong> user, then you can install ripgrep from the
<a href="http://openports.se/textproc/ripgrep">official ports</a>:</p>
<pre><code>$ doas pkg_add ripgrep
</code></pre>
<p>If you’re a <strong>NetBSD</strong> user, then you can install ripgrep from
<a href="http://pkgsrc.se/textproc/ripgrep">pkgsrc</a>:</p>
<pre><code># pkgin install ripgrep
</code></pre>
<p>If you’re a <strong>Haiku x86_64</strong> user, then you can install ripgrep from the
<a href="https://github.com/haikuports/haikuports/tree/master/sys-apps/ripgrep">official ports</a>:</p>
<pre><code>$ pkgman install ripgrep
</code></pre>
<p>If you’re a <strong>Haiku x86_gcc2</strong> user, then you can install ripgrep from the
same port as Haiku x86_64 using the x86 secondary architecture build:</p>
<pre><code>$ pkgman install ripgrep_x86
</code></pre>
<p>If you’re a <strong>Rust programmer</strong>, ripgrep can be installed with <code>cargo</code>.</p>
<ul>
<li>Note that the minimum supported version of Rust for ripgrep is <strong>1.34.0</strong>,
although ripgrep may work with older versions.</li>
<li>Note that the binary may be bigger than expected because it contains debug
symbols. This is intentional. To remove debug symbols and therefore reduce
the file size, run <code>strip</code> on the binary.</li>
</ul>
<pre><code>$ cargo install ripgrep
</code></pre>
<h3>Building</h3>
<p>ripgrep is written in Rust, so you’ll need to grab a
<a href="https://www.rust-lang.org/">Rust installation</a> in order to compile it.
ripgrep compiles with Rust 1.34.0 (stable) or newer. In general, ripgrep tracks
the latest stable release of the Rust compiler.</p>
<p>To build ripgrep:</p>
<pre><code>$ git clone https://github.com/BurntSushi/ripgrep
$ cd ripgrep
$ cargo build --release
$ ./target/release/rg --version
0.1.3
</code></pre>
<p>If you have a Rust nightly compiler and a recent Intel CPU, then you can enable
additional optional SIMD acceleration like so:</p>
<pre><code>RUSTFLAGS=&quot;-C target-cpu=native&quot; cargo build --release --features 'simd-accel'
</code></pre>
<p>The <code>simd-accel</code> feature enables SIMD support in certain ripgrep dependencies
(responsible for transcoding). They are not necessary to get SIMD optimizations
for search; those are enabled automatically. Hopefully, some day, the
<code>simd-accel</code> feature will similarly become unnecessary. <strong>WARNING:</strong> Currently,
enabling this option can increase compilation times dramatically.</p>
<p>Finally, optional PCRE2 support can be built with ripgrep by enabling the
<code>pcre2</code> feature:</p>
<pre><code>$ cargo build --release --features 'pcre2'
</code></pre>
<p>(Tip: use <code>--features 'pcre2 simd-accel'</code> to also include compile time SIMD
optimizations, which will only work with a nightly compiler.)</p>
<p>Enabling the PCRE2 feature works with a stable Rust compiler and will
attempt to automatically find and link with your system’s PCRE2 library via
<code>pkg-config</code>. If one doesn’t exist, then ripgrep will build PCRE2 from source
using your system’s C compiler and then statically link it into the final
executable. Static linking can be forced even when there is an available PCRE2
system library by either building ripgrep with the MUSL target or by setting
<code>PCRE2_SYS_STATIC=1</code>.</p>
<p>ripgrep can be built with the MUSL target on Linux by first installing the MUSL
library on your system (consult your friendly neighborhood package manager).
Then you just need to add MUSL support to your Rust toolchain and rebuild
ripgrep, which yields a fully static executable:</p>
<pre><code>$ rustup target add x86_64-unknown-linux-musl
$ cargo build --release --target x86_64-unknown-linux-musl
</code></pre>
<p>Applying the <code>--features</code> flag from above works as expected. If you want to
build a static executable with MUSL and with PCRE2, then you will need to have
<code>musl-gcc</code> installed, which might be in a separate package from the actual
MUSL library, depending on your Linux distribution.</p>
<h3>Running tests</h3>
<p>ripgrep is relatively well-tested, including both unit tests and integration
tests. To run the full test suite, use:</p>
<pre><code>$ cargo test --all
</code></pre>
<p>from the repository root.</p>
<h3>Translations</h3>
<p>The following is a list of known translations of ripgrep’s documentation. These
are unofficially maintained and may not be up to date.</p>
<ul>
<li><a href="https://github.com/chinanf-boy/ripgrep-zh#%E6%9B%B4%E6%96%B0-">Chinese</a></li>
</ul>

